README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# safar6

<!-- badges: start -->

[![Project Status: Concept – Minimal or no implementation has been done
yet, or the repository is only intended to be a limited example, demo,
or
proof-of-concept.](https://www.repostatus.org/badges/latest/concept.svg)](https://www.repostatus.org/#concept)
[![R-CMD-check](https://github.com/matt-dray/safar6/workflows/R-CMD-check/badge.svg)](https://github.com/matt-dray/safar6/actions)
[![CodeFactor](https://www.codefactor.io/repository/github/matt-dray/safar6/badge)](https://www.codefactor.io/repository/github/matt-dray/safar6)
[![rostrum.blog
post](https://img.shields.io/badge/rostrum.blog-post-008900?style=flat&labelColor=black&logo=data:image/gif;base64,R0lGODlhEAAQAPEAAAAAABWCBAAAAAAAACH5BAlkAAIAIf8LTkVUU0NBUEUyLjADAQAAACwAAAAAEAAQAAAC55QkISIiEoQQQgghRBBCiCAIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAAh+QQJZAACACwAAAAAEAAQAAAC55QkIiESIoQQQgghhAhCBCEIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAA7)](https://www.rostrum.blog/2021/01/04/safar6/)
<!-- badges: end -->

## Summary

An R package containing an [R6](https://r6.r-lib.org/) class to simulate
simplified gameplay elements of [the Safari
Zone](https://bulbapedia.bulbagarden.net/wiki/Kanto_Safari_Zone)
sub-area from [*Pokémon
Blue*](https://bulbapedia.bulbagarden.net/wiki/Generation_I) (1998) with
OOP.

The first generation of *Pokémon* games were developed for [the Nintendo
Game Boy](https://en.wikipedia.org/wiki/Game_Boy) by [Game
Freak](https://www.gamefreak.co.jp/) and published by
[Nintendo](https://www.nintendo.com). Pokémon as a property is owned by
[The Pokémon Company](https://www.pokemon.co.jp/).

This project is a demonstration of [the {R6}
package](https://r6.r-lib.org/) and the R6-class for object-oriented
programming in R. It’s purely for learning purposes. You can [read a
blog post](https://www.rostrum.blog/2021/01/04/safar6/) about it.

## Basic gameplay

Install from GitHub with `remotes::install_github("matt-dray/safar6")`.

The game is text-based and takes place entirely in the console. Start a
game like `x <- safar6::safari_zone$new()`. Be warned: it’s tricky. But
then so was the original.

Use `x$step()` to take a step through the tall grass of the Safari Zone.
There’s a chance you’ll encounter a wild Pokémon; perhaps it will be a
rare one at a high level. You’ll be prompted to throw a Safari Ball to
try and catch it, with the option to throw bait or a rock to influence
its catchability. Try to catch as many Pokémon as you can before you run
out of steps or balls.

Use `x$pause()` to ‘pause’ the game and see your current stats. You can
boot up `x$bills_pc` to see which Pokémon you’ve caught so far.

For help with fields and methods see `?safar6::safari_zone`.

<details>
<summary>
Expand for (contrived) example gameplay 👾
</summary>

``` r
> library(safar6)
# {safar6}
# Start game: x <- safari_zone$new()
# Take a step: x$step()
> x <- safari_zone$new()
# ------------------------
# First, what is your name?
# NEW NAME (1)
# BLUE (2)
# GARY (3)
# JOHN (4)
# ------------------------
# Select 1, 2, 3 or 4:
> 1
# Your name: 
> THEW
# Welcome to the SAFARI ZONE!
# For just P500, you can catch all the
# Pokemon you want in the park!
# Would you like to join the hunt?
# ------------------------
# MONEY: P500
# YES (1) or NO (2)
# ------------------------
# Select 1 or 2: 
> 1
# That'll be P500 please!
# ------------------------
# MONEY: P0
# ------------------------
# We only use a special POKe BALL here.
# THEW received 30 SAFARI BALLs!
# We'll call you on the PA when you run out of time or SAFARI BALLs!
> x$step()
# 499/500
> x$step()
# 498/500
> x$step()
# 497/500
# Wild VENONAT L22 appeared!
# ------------------------
# BALLx30 (1)     BAIT (2)
# THROW ROCK (3)  RUN (4)
# ------------------------
# Selection: 
> 3
# THEW threw a ROCK.
# Wild VENONAT is angry!
# ------------------------
# BALLx30 (1)     BAIT (2)
# THROW ROCK (3)  RUN (4)
# ------------------------
# Select 1, 2, 3 or 4: 
> 1
# THEW used SAFARI BALL!
# Wobble...
# Darn! The POKeMON broke free!
# Wild VENONAT is angry!
# ------------------------
# BALLx29 (1)     BAIT (2)
# THROW ROCK (3)  RUN (4)
# ------------------------
# Select 1, 2, 3 or 4: 
> 1
# THEW used SAFARI BALL!
# Wobble... Wobble... Wobble...
# All right!
# VENONAT was caught!
# Do you want to give a nickname to VENONAT?
# ------------------------
# YES (1) or NO (2)
# ------------------------
# Select 1 or 2:
> 1
# Nickname: 
> Tajiri
# Tajiri was transferred to BILL's PC!
> x$pause()
# 497/500
# BALLx28
# Transferred to BILL's PC: 1
> x$bills_pc
#   nickname species level
# 1   Tajiri VENONAT    22
```

</details>

## Graphics

Here’s a screenshot of the game in action using the
[gamelad](https://github.com/matt-dray/gamelad) RStudio theme and [the
VT323 font](https://fonts.google.com/specimen/VT323).

<img src="https://raw.githubusercontent.com/matt-dray/gamelad/main/img/rstudio-safar6-gamelad-vt323.png" alt="Screenshot of Rstudio showing only the console pane with green text and background, showing text that describes an encounter with a Pokemon." width="50%"/>

## Details

### The Safari Zone

The [Safari
Zone](https://bulbapedia.bulbagarden.net/wiki/Kanto_Safari_Zone) is an
enclosed area separate from the rest of the game. Game mechanics are
different here: you’re allowed 500 steps and 30 balls to capture
Pokémon. Also, wild encounters don’t result in a ‘battle’; instead you
can ‘bait’ or ‘throw a rock’ to influence the catch rate of the target.
I chose to simulate this area due to its relative simplicity compared to
other areas of the game, which use a more complex battle mechanic.

### Technicals

The fields in the `SafariZone` class are ‘overworld’ counters,
basically: the number of steps and balls remaining, and the number and
identity of captures you’ve made.

Methods are:

-   `$new()` to initialise the class and start the game
-   `$step()`, which moves your ‘player’ and may result in an encounter
-   `$pause()` (a copy of the `$print()` method) to simulate the pause
    screen from the game, showing the current number of steps and balls
    remaining

The `$step()` method does the hard work. It contains code that:

-   decides whether you encountered a target, based on the encounter
    rate of the Safari Zone (\~12%)
-   selects a wild Pokémon based on the relative encounter-rate slots
    for the Safari Zone Center Area particular to *Pokémon Blue*
    (e.g. 20% for a level 22 Nidoran Female through to 1% for a level 23
    Chansey; see `safar6::pokemon` for details)
-   generates individual variation in the selected Pokémon’s speed and
    HP (IV values), which influence catchability
-   prompts you interactively for an action: throw ball, bait, throw
    rock, run
-   adjusts the ‘angry’ and ‘eating’ values of the target and modifies
    catch rate and run probability as a result (a rock doubles the catch
    rate and the flee chance; bait halves the catch rate, but makes it
    four times less likely to run)
-   calculates whether the Pokémon will be caught or will run away
-   terminates the game if steps or balls are depleted

## Omissions

To keep things (much!) simpler than the original, this implementation
doesn’t include a number of features from the original game. Some
missing elements are:

-   visuals (everything is printed in the console)
-   movement around a map (all steps are treated as though you’re moving
    through tall grass)
-   all areas of the Safari Zone (only the Center area is used here)
-   finding items (they have no use in this context)
-   fishing and use of different rod types
-   species catch and encounter rates from the other Gen 1 versions
    (e.g. Scyther is in *Red* rather than Pinsir; and *Yellow* has some
    catch-rate changes)

## Sources

Information about game mechanics and values are relatively tricky to
come by. I used the following resources:

-   [Bulbapedia](https://bulbapedia.bulbagarden.net/) for Pokémon stats
    and game mechanics
-   [The Cave of Dragonflies](https://www.dragonflycave.com/) for game
    mechanics
-   the [Pokémon Slots](https://sites.google.com/site/pokemonslots)
    website for encounter rates
-   the [pret/pokered GitHub repo](https://github.com/pret/pokered) for
    a disassembly of the games

I also referred to a post I wrote about [using {R6} to simulate an
Automatic Bell
Dispenser](https://www.rostrum.blog/2020/04/04/repaying-tom-nook-with-r6/)
from Nintendo’s *Animal Crossing: New Horizons* (2020).

## Code of Conduct

Please note that the {safar6} project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## Disclaimer

I do not condone throwing rocks at any creature, let alone catching and
imprisoning them.


NEWS.md
# safar6 0.1.1

* Text is progressively revealed with `cat_tw()` (#28).

# safar6 0.1.0

* De-lint the package (#25).
* Add {lintr} GitHub Action.
* Bump to 0.1.0, tag and release.

# safar6 0.0.0.9007

* Added name input to initialise method (#7).
* Add `$name` field.

# safar6 0.0.0.9006

* Improved code comments.
* Added while loops to restrict `readline()` input (#13)
* Added `.onAttach()`/`zzz.R` for startup prompt on how to play.

# safar6 0.0.0.9005

* Fixed bug with game ending at 0 steps (#20).
* Fixed bug with Pokemon running away after another event has already occurred (#18).

# safar6 0.0.0.9004

* Added interactive nicknames (#17).

# safar6 0.0.0.9003

* Added wobble logic (#5).

# safar6 0.0.0.9002

* Added IV calculations (#8).
* Added end-game results (#4).
* Tidied while-loop to remove flow isses.

# safar6 0.0.0.9001

* Added `pokemon.rda` to `data/`.
* Added encounter while loop to the `SafariZone` class (#1).

# safar6 0.0.0.9000

* Added news, readme, license, code of conduct.
* Added `SafariZone` class.


R/data.R
#' Data for Encounters with Wild Pokemon
#'
#' A dataset containing the Pokemon you can encounter in the Safari Zone in
#' Pokemon Blue and their statistics.
#'
#' @format A data frame with 10 rows and 10 variables:
#' \describe{
#'   \item{game}{The Pokemon generation 1 game variant. \code{BLUE} only for
#'     now.}
#'   \item{area}{The area of the Safari Zone (\code{CENTER} only for now)}
#'   \item{tile}{The terrain on which the player is standing (\code{GRASS} only
#'     for now).}
#'   \item{species}{The encounter Pokemon's species.}
#'   \item{level}{The encounter Pokemon's level.}
#'   \item{hp_base}{The base speed for the species.}
#'   \item{speed_base}{The base speed for the species.}
#'   \item{catch_base}{The base catch rate for the species.}
#'   \item{slot}{Encounter rate slot, which determines \code{encounter_rate}}
#'   \item{encounter_rate}{Chance of encounter.}
#' }
#' @source \url{https://github.com/pret/pokered} and
#'     \url{https://bulbapedia.bulbagarden.net/}
"pokemon"
#> [1] "pokemon"


R/class.R

#' R6 Class Representing the Safari Zone from Pokemon Blue (1998)
#'
#' @description
#' An \code{\link[R6]{R6Class}} object to simulate simplified gameplay elements
#' of the Safari Zone sub-area from \emph{Pokemon Blue} (1998).
#'
#' @details
#' The first generation of \emph{Pokemon} games were developed for the Nintendo
#' Game Boy by Game Freak (\url{https://www.gamefreak.co.jp/}) and published by
#' Nintendo (\url{https://www.nintendo.com}). Pokemon as a property is owned by
#' The Pokemon Company (\url{https://www.pokemon.co.jp/}).
#'
#' @export
#'
#' @examples \dontrun{ x <- safar6::safari_zone$new() }
safari_zone <- R6::R6Class("SafariZone",

  public = list(

    # Fields ----

    #' @field name Character. Player's name.
    name = "BLUE",
    #' @field steps Numeric. Steps remaining (500 at start).
    steps = 500,
    #' @field balls Numeric. Safari Balls remaining (30 at start).
    balls = 30,
    #' @field captures Numeric. Count of wild Pokemon captured (0 at start).
    captures = 0,
    #' @field bills_pc Dataframe. Details of wild Pokemon caught (empty at
    #'     start).
    bills_pc = data.frame(nickname = NULL, species = NULL, level = NULL),

    # Methods ----

    #' @description Create a new Safari Zone object.
    #' @return A \code{SafariZone}-class object.
    initialize = function() {

      # Ask for player name
      cat_tw("First, what is your name?\n")
      cat(
        "------------------------\n",
        "NEW NAME (1)\nBLUE (2)\nGARY (3)\nJOHN (4)\n",
        "------------------------\n",
        sep = ""
      )

      # Restrict answer possibilities
      response_name <- 0  # set variable outside 1 to 4 to start
      while (!response_name %in% 1:4) {
        response_name <- readline("Select 1, 2, 3 or 4: ")
      }

      # Set player name
      if (response_name == "1") {
        response_name <- readline("Your name: ")  # ask interactively for name
        self$name <- response_name
      } else if (response_name == "2") {
        self$name <- "BLUE"
      } else if (response_name == "3") {
        self$name <- "GARY"
      } else if (response_name == "4") {
        self$name <- "JOHN"
      }

      # Welcome message
      cat_tw(c(
        "Welcome to the SAFARI ZONE!\n",
        "For just P500, you can catch all the\n",
        "Pokemon you want in the park!\n",
        "Would you like to join the hunt?\n"
      ))
      cat(
        "------------------------\n",
        "MONEY: P500\n",
        "YES (1) or NO (2)\n",
        "------------------------\n",
        sep = ""
      )

      # Restrict answer possibilities
      response_pay <- 0  # set variable outside 1 or 2 to start
      while (!response_pay %in% 1:2) {
        response_pay <- readline("Select 1 or 2: ")
      }

      if (response_pay == "1") {

        # Instructions to player
        cat_tw("That'll be P500 please!\n")
        cat(
          "------------------------\n",
          "MONEY: P0\n",
          "------------------------\n",
          sep = ""
        )
        cat_tw(c(
          "We only use a special POKe BALL here.\n",
          self$name, " received 30 SAFARI BALLs!\n",
          "We'll call you on the PA when you run\n",
          "out of time or SAFARI BALLs!\n"
        ))

      } else if (response_pay == "2") {

        # Lol, you do want to play!
        cat_tw("...", 1)
        cat_tw("That'll be P500 please!\n")
        cat(
          "------------------------\n",
          "MONEY: P0\n",
          "------------------------\n",
          sep = ""
        )
        cat_tw(c(
          "We only use a special POKe BALL here.\n",
          self$name, " received 30 SAFARI BALLs!\n",
          "We'll call you on the PA when you run\n",
          "out of time or SAFARI BALLs!\n"
        ))

      }

    },

    #' @description
    #' Create a new Safari Zone print method.
    #' @return A console message with steps and balls remaining.
    #' @examples \dontrun{
    #'     x <- safar6::safari_zone$new()  # initialise class
    #'     x$print()  # print the object, see stats
    #' }
    print = function() {
      cat(
        paste0(self$steps, "/500\n"),
        paste0("BALLx", self$balls, "\n"),
        paste0("BILL's PC: ", self$captures, "\n"),
        sep = ""
      )
    },

    #' @description
    #' Simulate the pause function from the original game.
    #' @return A console message with steps and balls remaining.
    #' @seealso \code{\link{print}}
    #' @examples \dontrun{
    #'     x <- safar6::safari_zone$new()  # intialise class
    #'     x$pause()  # 'pause' the game, see stats
    #' }
    pause = function() {
      self$print()
    },

    #' @description
    #' Take a step in the Safari Zone.
    #' @return Either nothing, or a wild encounter.
    #' @examples \dontrun{
    #'     x <- safar6::safari_zone$new()  # initialise class
    #'     x$step()   # take step, prints steps remaining
    #' }
    step = function() {

      # Adjust and check steps ----

      # Reduce step count field by 1
      if (self$steps > 0) {
        self$steps <- self$steps - 1
        cat(paste0(self$steps, "/500\n"))  # let player know steps remaining
      }

      # Check if steps have run out, end game if so
      if (self$steps == 0) {

        # End-game notice and results
        cat_tw(c(
          "PA: Ding-dong!\nTime's up!\nPA: Your SAFARI GAME is over!\n",
          "Did you get a good haul?\nCome again!\n"
        ))
        cat(
          "------------------------\nResult: ",
          self$captures, " transferred to BILL's PC\n",
          sep = ""
        )

        # Print details of any catches
        if (self$captures > 0) print(self$bills_pc)

      }

      # Begin encounter ----

      if (sample(0:255, 1) < 30 &  # Safari Zone base encounter rate is 30
          self$steps > 0 & self$balls > 0) {  # balls/steps must be non-zero

        # The encounter has begun, set variable
        # The while-Loop of turns is broken when this gets modified to FALSE
        encounter_active <- TRUE

        # Wild Pokemon selection, IV calculation ----

        # Select a random species/level weighted by encounter rate
        pkmn <-
          safar6::pokemon[sample(seq_along(safar6::pokemon), 1,
                                 prob = safar6::pokemon$encounter_rate), ]

        # Print notice of the wild Pokemon selected
        cat_tw(c(
          "Wild ", pkmn$species, paste0(" L", pkmn$level), " appeared!\n"
        ))

        # Set individual values (IVs)
        iv_atk <- sample(1:15, 1)  # attack
        iv_def <- sample(1:15, 1)  # defence
        iv_spd <- sample(1:15, 1)  # speed
        iv_spc <- sample(1:15, 1)  # 'special'

        # The HP IV depends on the other IVs and whether they're odd
        iv_hp <- 0
        if (iv_atk %% 2 != 0) iv_hp <- iv_hp + 8
        if (iv_def %% 2 != 0) iv_hp <- iv_hp + 4
        if (iv_spd %% 2 != 0) iv_hp <- iv_hp + 2
        if (iv_spc %% 2 != 0) iv_hp <- iv_hp + 1

        # Calculate the individual HP and speed for the wild Pokemon
        pkmn$hp_indiv <- floor(
          (((pkmn$hp_base + iv_hp) * 2) * pkmn$level) / 100
        ) + pkmn$level + 10
        pkmn$speed_indiv <-
          floor((((pkmn$speed_base + iv_spd) * 2) * pkmn$level) / 100) + 5

        # Set variables for modifiable statuses that impact catch chance
        status_eating <- 0
        status_angry  <- 0
        status_catch <- pkmn$catch_base

        # Start the turns of the encounter ----

        # Turns loop from player to Pokemon until encounter_active is FALSE
        while (encounter_active == TRUE & self$steps > 0) {

          # Adjust eating/anger status ----

          # Pokemon's accrued anger/eating values depleted by 1 if non-zero
          if (status_eating > 0) {
            cat_tw(c("Wild ", pkmn$species, " is eating!\n"))
            status_eating <- status_eating - 1
          } else if (status_angry > 0) {
            cat_tw(c("Wild ", pkmn$species, " is angry!\n"))
            status_angry <- status_angry - 1
            if (status_angry == 0) {
              status_catch <- pkmn$catch_base  # returns to original rate
            }
          }

          # Player's turn ----

          # Ask player to choose from options
          cat(
            "------------------------\n",
            paste0(
              "BALLx",
              formatC(self$balls, width = 2, format = "d", flag = "0")
            ),
            " (1)     BAIT (2)\nTHROW ROCK (3)  RUN (4)\n",
            "------------------------\n",
            sep = ""
          )

          # Collect player's selection, re-ask if not 1 to 4
          response_action <- 0  # set variable outside 1 to 4 to start
          while (!response_action %in% 1:4) {
            response_action <- readline("Select 1, 2, 3 or 4: ")
          }

          # Ifelse chain dependent on player's choice
          if (response_action == "1") {  # throw Safari Ball

            # Chose to throw ball ----
            cat_tw(c(self$name, " used SAFARI BALL!\n"))

            # Reduce ball count by 1
            self$balls <- self$balls - 1

            # Define HP-related catch factor, 'F'
            f1 <- (pkmn$hp_indiv * 255) / 12  # 12 is Safari Ball factor
            f2 <- max(pkmn$hp_indiv / 4, 1)  # divide by 4, but can't be 0
            f3 <- min(f1 / f2, 255)  # divide but can't be >255

            # Define wobble factor, 'w'
            w <- (floor((pkmn$catch_base * 100) / 150) * f3) / 255  # wobble
            s <- 1  # wobble delay in seconds

            # Generate RNG for ball and HP
            ball_rng <- sample(0:150, 1)  # 150 specific to Safari Ball
            hp_rng <- sample(0:255, 1)

            # Ifelse for capture
            if (status_catch < ball_rng) {  # ball-related catch fail

              # Breakout
              cat_tw("Wobble...\n")
              Sys.sleep(s)
              cat_tw("Darn! The POKeMON broke free!\n")

            } else if (status_catch >= ball_rng) {  # ball-related catch success

              if (f3 >= hp_rng) {  # capture

                # Wobble
                cat_tw("Wobble... ")
                Sys.sleep(s)
                cat_tw("Wobble... ")
                Sys.sleep(s)
                cat_tw("Wobble...\n")
                Sys.sleep(s)

                # Print capture notice, ask for nickname
                cat_tw(c(
                  "All right!\n", pkmn$species, " was caught!\n",
                  "Do you want to give a nickname to ", pkmn$species, "?\n"
                ))
                cat(
                  "------------------------\n",
                  "YES (1) or NO (2)\n",
                  "------------------------\n",
                  sep = ""
                )

                # Collect player's response, must be 1 or 2
                response_nickname <- 0  # set variable outside 1 or 2 to start
                while (!response_nickname %in% 1:2) {
                  response_nickname <- readline("Select 1 or 2: ")
                }

                # Prompt for nickname if yes, otherwise nickname is species name
                if (response_nickname == "2") {
                  pkmn$nickname <- pkmn$species  # assign species to nickname
                  cat_tw(c(pkmn$species, " was transferred to BILL's PC!\n"))
                } else if (response_nickname == "1") {
                  response_nickname <- readline("Nickname: ")  # ask for name
                  pkmn$nickname <- response_nickname  # assign to wild Pokemon
                  cat_tw(c(
                    response_nickname, " was transferred to BILL's PC!\n"
                  ))
                }

                # Incrementcapture counter by 1
                self$captures <- self$captures + 1

                # Append captured Pokemon information to Bill's PC field
                self$bills_pc <- rbind(
                  self$bills_pc,
                  data.frame(
                    nickname = pkmn$nickname,
                    species = pkmn$species,
                    level = pkmn$level
                  )
                )

                # Break loop, Pokemon was captured
                encounter_active <- FALSE

              } else {  # second break-free chance, based on wobble factor

                # HP always max in Safari Zone, so multi-wobble unlikely
                if (w < 10) {
                  cat_tw("The ball missed the POKeMON!\n")
                } else if (w %in% 10:29) {
                  cat_tw("Wobble...\n")
                  Sys.sleep(s)
                  cat_tw("Darn! The POKeMON broke free!\n")
                } else if (w %in% 30:69) {
                  cat_tw("Wobble... ")
                  Sys.sleep(s)
                  cat_tw("Wobble...\n")
                  Sys.sleep(s)
                  cat_tw("Aww! It appeared to be caught!\n")
                } else if (w > 70) {
                  cat_tw("Wobble... ")
                  Sys.sleep(s)
                  cat_tw("Wobble... ")
                  Sys.sleep(s)
                  cat_tw("Wobble...\n")
                  Sys.sleep(s)
                  cat_tw("Shoot! It was so close too!\n")
                }

              }

            }

          } else if (response_action == "2") {  # throw bait

            # Chose to throw bait ----
            cat_tw(c(self$name, " threw some BAIT.\n"))

            # Make status adjustments
            status_catch  <- floor(status_catch / 2)  # halve catch rate
            status_eating <- status_eating + sample(1:5, 1)  # adjust +1 to 5
            status_eating <- min(status_eating, 255) # cap at 255
            status_angry  <- 0  # reset anger status

          } else if (response_action == "3") {  # throw rock

            # Chose to throw a rock ----
            cat_tw(c(self$name, " threw a ROCK.\n"))

            # Make status adjustments
            status_catch  <- min(status_catch * 2, 255)  # double catch rate
            status_angry  <- status_angry + sample(1:5, 1)  # adjust +1 to 5
            status_angry  <- min(status_angry, 255) # cap at 255
            status_eating <- 0  # reset eating status

          } else if (response_action == "4") {  # exit encounter

            # Chose to run ----
            cat_tw("Got away safely!\n")

            # Break loop, player has voluntarily left the encounter
            encounter_active <- FALSE

          }

          # Wild Pokemon's turn ----

          # Define chance to run away, 'x', based on individual speed
          x <- (pkmn$speed_indiv %% 256) * 2

          # Adjust run chance based on eating/anger status
          if (status_eating > 0) {
            x <- x / 4  # flee chance reduced to a quarter
          } else if (status_angry > 0) {
            x <- min(x * 2, 255)  # flee chance doubled
          }

          # Set flee RNG
          flee_rng <- sample(0:255, 1)

          if (encounter_active == TRUE & flee_rng < x) {
            cat_tw(c("Wild ", pkmn$species, " ran away!\n"))
            encounter_active <- FALSE  # break loop, Pokemon has fled
          }

          # Check ball counter ----

          # Check if balls remain
          if (self$balls == 0) {

            # End game notice and results
            cat_tw(c(
              "PA: Ding-dong!\nTime's up!\nPA: Your SAFARI GAME is over!\n",
              "Did you get a good haul?\nCome again!\n"
            ))
            cat(
              "------------------------\nResult: ",
              self$captures, " transferred to BILL's PC\n",
              sep = ""
            )

            # Print details of any catches
            if (self$captures > 0) print(self$bills_pc)

            # Game is over, no more steps can be taken, set field to 0
            self$steps <- 0

            # Break loop, player is out of balls and the game is over
            encounter_active <- FALSE

          }

        }

      }

      invisible(self)

    }

  )

)
